[Sean Ng] iP

snss231 opened the pull request:
# Duke bro

- this
- is
- a
- bullet
- list

1. this
2. is
3. a
4. numbered
5. list

```java
public static void Main(String[] args) {
   System.out.println("This is a fenced code block");
}
```
- [ ] this
- [ ] is
- [ ] a
- [ ] task
- [x] list

👍 Emojis are cool

> "Do. Or do not. There is no try." - Yoda

[This is a hyperlink](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

This is `inline code`.

**bold** *italic* ~~strikethrough~~

snss231 committed:
Level-1

snss231 committed:
Level-2

snss231 committed:
Level-3 A-Classes

snss231 committed:
Level-4 A-Inheritance

snss231 committed:
A-TextUiTesting

snss231 committed:
Level 5 A-Exceptions

snss231 committed:
Level-6 A-Collections

snss231 committed:
A-Enums

snss231 committed:
unindent switch cases

snss231 committed:
Add save functionality, add more exceptions

snss231 committed:
Handle data folder does not exist

snss231 committed:
Add basic date functionality for Deadline

snss231 committed:
Merge branch 'branch-level-7'

snss231 committed:
Merge branch 'branch-level-8'

snss231 committed:
Implement TaskList, Storage, add data/ to .gitignore

snss231 committed:
Create and implement Command class + subclasses

snss231 committed:
Create and implement Command class + subclasses

snss231 committed:
Create and implement Command + subclasses

snss231 committed:
Implement ui methods, bro-ify messages, organise into packages"

snss231 committed:
Add unit tests for Deadline and Todo

snss231 committed:
Add javadocs for commands

snss231 committed:
Add javadocs for tasks

snss231 committed:
Add javadocs for Duke, DukeException, Parser, Storage, TaskList, Ui

snss231 committed:
Align to coding standard

snss231 committed:
Add find functionality

snss231 committed:
Fix merge conflicts

snss231 committed:
Fix merge conflicts

snss231 committed:
Fix typo in showMarkTask header

aweijun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,100 @@
+     * @throws DukeException If user's input is invalid (does not correspond to any command or invalid parameters).
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] input = fullCommand.split(" ", 2); //return [commandWord, restOfCommand]
```
Should the variable be plural instead?

input seems to be an array, according to the coding standards it should be plural.

aweijun commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Storage {
+
+    static final String PROJECT_ROOT = System.getProperty("user.dir");
```
Should there be an access modifier here?

There seems to be a missing access modifier.

Considering adding a private statement here.

aweijun commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,118 @@
+        System.out.println("Bye bro, see you again soon.");
+    }
+
+    /**
+     * Prints all the tasks in the list of tasks.
+     *
+     * @param tasks The list of tasks.
+     */
+    public void showTasks(TaskList tasks) {
+        System.out.println(
+                tasks.isEmpty()
+                ? "There are no tasks yet, bro."
+                : "Here are your tasks bro:\n" + tasks);
+    }
+
+    /**
+     * Notifies the user upon task successfully marked.
+     *
+     * @param task The marked task.
+     */
+    public void showMarkTask(Task task) {
+        System.out.println("Ok bro, marked this task:\n" + task);
+    }
+
+    /**
+     * Notifies the user upon task successfully unmarked.
+     *
+     * @param task The unmarked task.
+     */
+    public void showUnmarkTask(Task task) {
+        System.out.println("Ok bro, unmarked this task:\n" + task);
+    }
+
+    public void showFindTasks(TaskList results, String keyword) {
+        System.out.println(
+                results.isEmpty()
+                        ? "There are no tasks with the keyword \"" + keyword + "\" bro."
+                        : "Here are the matching tasks in your list:\n" + results);
+    }
+}
```
Nice abstraction! I should consider doing it for mine as well.

aweijun commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,25 @@
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + at + ")";
+    }
+}
```
Should there be a new line here?

Considering adding a new line here, as it would be flagged up during the check style.

aweijun commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+     * @return The completed task.
+     */
+    public Task mark(int index) {
+        Task task = this.tasks.get(index);
```
I am unsure if "this" statement is necessary?

This doesn't recommend based on the coding standard.

Can considering removing this statement.

snss231 replied:
I wasn't aware of this, my perception was that the "this" keyword should be used where possible to make the code more explicit. I tried finding information related to the "this" keyword in the coding standard document @ https://se-education.org/guides/conventions/java/intermediate.html but couldn't find it, mind pointing me in the right direction? thanks 😄

aweijun reviewed (commented):
Looks good, learned a lot from reviewing your IP!
Some cosmetic changes that were proposed.

aweijun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,100 @@
+package duke;
+
+import duke.command.*;
```
Shouldn't import statements be explicitly stated instead?

Consider listing out each one explicitly.

snss231 replied:
I agree with all your comments, thanks for helping me spot errors wrt coding standard. Cheers!

aweijun reviewed (commented):
Great work for the IP!

BrendonLau commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * Commands do not cause program to exit by default.
+     */
+    public boolean isExit() {
```
So far good job in the naming convention.

snss231 replied:
Thanks! All the best for your iP 😄

BrendonLau reviewed (commented):
Other than what the other reviewer mentioned. LGTM. Good job!

snss231 committed:
Update README.md

snss231 committed:
Fix minor coding standard issues

snss231 committed:
Fix import *

snss231 committed:
Add gradle

snss231 committed:
Merge

snss231 committed:
Add checkstyle and fix coding violations

snss231 committed:
Implement GUI

snss231 committed:
Add bot logic to GUI

clement0010 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+        return isDone ? "X" : " ";
+    }
+
+    public boolean descriptionContains(String keyword) {
```
Perhaps `containsDescription` or `hasDescription` would sound better for this function?

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+                return new MarkCommand(index);
+            } catch (NumberFormatException e) {
+                throw new DukeException("Invalid parameter(s). Usage: mark [taskNumber]");
+            } catch (IndexOutOfBoundsException e) {
+                throw new DukeException("Invalid parameter(s). duke.task.Task " + (index + 1) + " does not exist");
```
I think `IndexOutOfBoundsException` will never be caught here? The error will be thrown in the `execute` function and the same error was handled there as well.

clement0010 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public TaskList load() {
+        Path dataFolder = Path.of(PROJECT_ROOT, "data");
+        if (!Files.exists(dataFolder)) { //initialize data directory if not present
```
Good use of guard clause here

clement0010 reviewed (commented):
Overall, the code follows coding standards and not many issues to nitpick. Good Job!

snss231 committed:
Add assertions in Duke constructor

snss231 committed:
Split up ShowFindTasks and ShowTasks to improve code quality

snss231 committed:
Merge pull request #2 from snss231/branch-A-Assertions

Add assertions in Duke constructor to ensure fields are properly initialized.

snss231 committed:
Merge branch 'master' into branch-A-CodeQuality

snss231 committed:
Merge pull request #3 from snss231/branch-A-CodeQuality

Split up ShowFindTasks and ShowTasks to improve code quality

snss231 committed:
Setup Github Actions CI

snss231 committed:
Edit workflow file

snss231 committed:
Remove unused imports

snss231 committed:
There is no help command.

Users have no way to list all commands.

Let's insert a help command so that users can see all commands and their
syntax and purpose.

snss231 committed:
Add background, adjust padding

snss231 committed:
Add screenshot

snss231 committed:
Make photos circle

snss231 committed:
Update ss

snss231 committed:
Update ss

snss231 committed:
Update README.md

snss231 committed:
Update README.md

snss231 committed:
Update README.md

snss231 committed:
Update README.md

snss231 committed:
Update README.md

snss231 committed:
Update README.md

snss231 committed:
Fix checkstyle
